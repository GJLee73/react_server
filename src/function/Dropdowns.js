import React, { Component } from 'react';
import styled from 'styled-components'
class Dropdowns extends Component {
    
      constructor(props) {
        super(props);
        this.state = {  }
      }
      render() { 
        return ( 
          <div className={this.props.className}>
        <Dropdown>
          <div className='title' onClick={this.props.toggle}>{this.props.title}  <span></span></div> 
          {this.props.open?
          
            <ClickOutside toggle={this.props.toggle}>
              {this.props.children}
            </ClickOutside>

          :undefined
        }
          
        </Dropdown>
        </div>
      )}
    
}
const Dropdown = styled.div`
  span{
    display:inline-block; width:0; height:0; border-style:solid; border-width:4px;
    border-color:#56a7ff transparent transparent transparent;

  }
 .title{
   cursor:pointer;
 }
 .dropdownMenu{
   position:absolute;
   width:100px;
   background-color:white;
   border:solid 1px #acacac;
   border-radius:5px;
 }
`
 
export default Dropdowns;

class ClickOutside extends Component{
    constructor(props) {
      super(props);
      this.state = {  }
      this.setWrapperRef = this.setWrapperRef.bind(this);
      this.handleClickOutside = this.handleClickOutside.bind(this);
      this.wrapperRef=React.createRef();
  }
  componentDidMount() {
      document.addEventListener('mousedown', this.handleClickOutside);
    }

  componentWillUnmount() {
      document.removeEventListener('mousedown', this.handleClickOutside);
  }
  /**
  * Set the wrapper ref
  */
  setWrapperRef(node) {
  this.wrapperRef = node;
  }

  /**
  * Alert if clicked on outside of element
  */
  handleClickOutside(event) {
  if (this.wrapperRef && !this.wrapperRef.contains(event.target)&& event.target.className!=='title') {
    this.props.toggle()
  }
  }
  render() { 
      return ( <div ref={this.setWrapperRef} className='dropdownMenu'>{this.props.children}</div> )
  }
}