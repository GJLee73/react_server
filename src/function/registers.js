
export function autoHypenPhone(phone){
    var str = phone.replace(/[^0-9]/g, '')
    
    var tmp = '';
    if( str.length < 4){
      return tmp;
    }else if(str.length < 7){
      tmp += str.substr(0, 3);
      tmp += '-';
      tmp += str.substr(3);
      return tmp
    }else if(str.length < 11){
      tmp += str.substr(0, 3);
      tmp += '-';
      tmp += str.substr(3, 3);
      tmp += '-';
      tmp += str.substr(6);
      return tmp
    }else{        
      tmp += str.substr(0, 3);
      tmp += '-';
      tmp += str.substr(3, 4);
      tmp += '-';
      tmp += str.substr(7);
      return tmp
    }
  }

export function  verifyPwd(pwd){
    var stringRegx = /[!@#$%^&*]/gi; 
    if(pwd.length>2){
        if(!stringRegx.test(pwd)){
            return {pwdVerifyPhrase:'최소 한개 이상의 특수문자가 포함되어야 합니다.(!@#$%^&*)'}
        }else if(pwd.length<8){
            return {pwdVerifyPhrase:'비밀번호가 너무 짧습니다.'}
        }
       return {pwdVerifyPhrase:'안전한 비밀번호입니다.',pwdVerified:true}
    }  
   
}
export function verifyEmail(email){
    var reg=/^[0-9a-zA-Z]([-_]?[0-9a-zA-Z])*@[0-9a-zA-Z]([-_.]?[0-9a-zA-Z])*.[a-zA-Z]{2,3}$/
    if(!reg.test(email)){
        alert('올바른 이메일을 입력하세요!')
        return 0;
    }
    console.log(2)
    return(true)
}
  