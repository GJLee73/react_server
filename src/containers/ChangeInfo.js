/**
 * 실제로 정보를 바꾸는 함수.
 * @author G1
 * @logs // 18.2.27
 */

import React from 'react';
import {connect} from 'react-redux'
import {getAllInfo} from '../modules/authentication'
import {newRegister, emailRequest, phoneRequest, nicknameRequest} from'../modules/register'
import {ChangeInfoPwd} from '../components/auth'
import {ChangeInfoComponent} from '../components/auth';
import {verifyEmail, autoHypenPhone} from '../function/registers'

export class ChangeInfo extends React.Component {
    constructor(props) {
        super(props);
        this.state = {typeName :this.props.match.params.typename,value:'' }
       this.submit=this.submit.bind(this)
    }
    /**
     * nickname이면 바로 정보 수정.
     * 다른 애들은 비밀번호 체크가 되어있는지를 확인
     * pwd면 정보를 보여주지 않고 0을 리턴
     */
    componentWillMount(){
       
        if(this.state.typeName!=='nickname'){
            
            if(!this.props.pwdVerified){
                alert('비밀번호 체크를 먼저 해주세요!')
                this.props.history.push('/settings')
                return 0;
                
            }
        }
        if(this.state.typeName=='password'){
            return 0;
        }
        this.setState({value:this.props.allInfo.get(this.state.typeName)})
       
    }



    /**
     * 개인정보를 변경하는 함수.
     * db에 등록 후 세팅화면으로 리다이렉트
     * @param {object} msg - 바꾸려고 하는 정보 인포.{typename:'string'}
     * check가 되어있지 않다면(서버에서 통과하지 못하면)return 0
     */
    async submit(msg){
        if(this.state.typeName!=='password')
            switch(this.state.typeName){
                case 'nickname': 
                await this.props.nicknameRequest(msg.nickname).catch(err=>console.log(err))
                if(!this.props.nicknameCheck){
                    alert('이미 등록된 닉네임입니다.')
                    return 0
                }
                break;
                case 'email':
                if(!verifyEmail(msg.email)){
                    return 0
                }
                //TODO: catch로 받는지, then으로 받는지 라우터 할때 같이 해주어야함
                await this.props.emailRequest(msg.email).catch((err)=>{console.log(err)})
   
                if(!this.props.emailCheck){
                    alert('이미 등록된 이메일입니다.')
                    return 0
                }
                break;
                
                case 'phone':
                this.props.phoneRequest(msg.phone)
                if(!this.props.phoneCheck){
                    alert('이미 등록된 번호입니다.')
                    return 0
                }
                break
            }
        
          this.props.newRegister(msg,this.props.status.get('currentUser')).then(()=>{
            alert('변경되었습니다.')
            this.props.history.push('/settings')
           }
           ).catch((err)=>{
               console.log(err)
           })
       
    }
    render() { 
        
        return ( 
        <div>
            새로운 {this.state.typeName}을 입력하세요.
            <div className='card container'>
            {this.state.typeName==='password'? <ChangeInfoPwd submit={this.submit}/>
             :<ChangeInfoComponent submit={this.submit} value={this.state.value} typeName={this.state.typeName}/>}
            
            
            </div>
        </div> )
    }
}
 

const mapStateToProps = (state) =>{
	return{
        allInfo: state.authentication.get('allInfo'),
        pwdVerified: state.authentication.get('pwdVerified'),
        emailCheck:state.register.emailCheck,
        phoneCheck:state.register.phoneCheck,
        nicknameCheck:state.register.nicknameCheck,
        status:state.authentication.get('status')
	};
};

const mapDispatchtoProps = (dispatch) => {
	return {
		
		getAllInfo:()=>{
            return dispatch(getAllInfo())
        },
        newRegister:(msg,id)=>{
            return dispatch(newRegister(msg,id))
        },
		emailRequest: (email)=>{
			return dispatch(emailRequest(email))
        },
        phoneRequest:(phone)=>{
            return dispatch(phoneRequest(phone))
        },
        nicknameRequest:(nickname)=>{
            return dispatch(nicknameRequest(nickname))
        }
       
	};
}

export default connect(mapStateToProps, mapDispatchtoProps)(ChangeInfo);