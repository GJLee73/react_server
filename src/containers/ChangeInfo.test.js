import React from 'react';
import {shallow, mount} from 'enzyme'
import {ChangeInfo} from './ChangeInfo'
import {Map} from 'immutable'
import {verifyEmail} from'../function/registers'

describe('ChangeInfo test',()=>{
    var submit = jest.spyOn(ChangeInfo.prototype,'submit')
    var getAllInfo=jest.fn(()=>Promise.resolve())
    var newRegister=jest.fn(()=>Promise.resolve())
    var allInfo=Map({nickname:'g1',email:'g1@naver',phone:'000-0000'})
    var history ={push:jest.fn()}
    global.alert=jest.fn()

    var wrapper = shallow(<ChangeInfo pwdVerified={true} match={{params:{typename:'email'}}}getAllInfo={getAllInfo} allInfo={allInfo} history={history}/>)

    describe('willmount test',()=>{
        it('typeName!==nickname, !==password, pwdVerified=true',()=>{
            expect(wrapper.state('typeName')).toBe('email')
            expect(wrapper.state('value')).toBe('g1@naver')
        })
        it('typeName!==nickname, ==password, pwdVerified=true',()=>{
            wrapper.setState({typeName:'password',value:''})
            wrapper.instance().componentWillMount()
            expect(wrapper.state('typeName')).toBe('password')
            expect(wrapper.state('value')).toBe('')
        })
        it('typeName!==nickname, pwdVerified=false',()=>{
            wrapper.setState({typeName:'phone',value:''})
            wrapper.setProps({pwdVerified:false})
            wrapper.instance().componentWillMount()
            expect(wrapper.state('typeName')).toBe('phone')
            expect(wrapper.state('value')).toBe('')
            expect(history.push.mock.calls.length).toBe(1)
            expect(global.alert.mock.calls[0][0]).toBe('비밀번호 체크를 먼저 해주세요!')
        })
        it('typeName==nickname',()=>{
            wrapper.setState({typeName:'nickname',value:''})
            wrapper.instance().componentWillMount()

            expect(wrapper.state('typeName')).toBe('nickname')
            expect(wrapper.state('value')).toBe('g1')
        })
    })

    describe('not password',()=>{
        describe('function test',()=>{
            describe('submit test',()=>{
                describe('typename:nickname',()=>{
                    it('nicknameCheck==true',async()=>{
                        wrapper.setProps({nicknameRequest:jest.fn(()=>Promise.resolve()),newRegister,nicknameCheck:true,status:Map({currentUser:'g1'})})
                        wrapper.setState({typeName:'nickname'})
                       await wrapper.instance().submit({nickname:'g1'}).catch((err)=>{console.log(err)})
                        expect(global.alert.mock.calls[1][0]).toBe('변경되었습니다.')
                    })
                    it('nicknameCheck==false',async()=>{
                        wrapper.setProps({nicknameCheck:false})
                        await wrapper.instance().submit({}).catch(err=>console.log(err))
                        expect(global.alert.mock.calls[2][0]).toBe('이미 등록된 닉네임입니다.')
                    })
                })
                describe('typename:email',()=>{
                    var emailRequest = jest.fn(()=>{return Promise.resolve()})
                    it('verifyEmail return false',()=>{
                        wrapper.setProps({emailRequest,emailRequest:()=>{return Promise.resolve()}})
                        wrapper.setState({typeName:'email'})
                        wrapper.instance().submit({email:'123'})
                        expect(global.alert.mock.calls[3][0]).toBe('올바른 이메일을 입력하세요!')
                        expect(emailRequest).toHaveBeenCalledTimes(0)
                    })
                    
                    it('emailCheck==false',async()=>{
                        wrapper.setProps({emailCheck:false})
                        await wrapper.instance().submit({email:'1234@naver.com'}).catch(()=>{console.log('4이 이상하다')})
                        expect(global.alert.mock.calls[4][0]).toBe('이미 등록된 이메일입니다.')
                        
                    })
                    it('emailCheck==true',async()=>{
                        wrapper.setProps({emailCheck:true})
                        await wrapper.instance().submit({email:'123@naver.com'}).catch(()=>{console.log('3이 이상하다')})
                        expect(global.alert.mock.calls[5][0]).toBe('변경되었습니다.')
                    })
                })
                describe('typename:phone',()=>{
                    it('phonecheck ==false',async()=>{
                        wrapper.setState({typeName:'phone'})
                        wrapper.setProps({phoneRequest:jest.fn(),phoneCheck:false})
                        await wrapper.instance().submit({phone:'000-0000-0000'}).catch(()=>{console.log('2이 이상하다')})
                        expect(global.alert.mock.calls[6][0]).toBe('이미 등록된 번호입니다.')
                    })
                    it('phonecheck==true',async()=>{
                        wrapper.setProps({phoneCheck:true})
                        await wrapper.instance().submit({phone:'012-3345-6666'}).catch(()=>{console.log('1이 이상하다')})
                        expect(global.alert.mock.calls[7][0]).toBe('변경되었습니다.')
                    })
                   
                })
             })
        })
    })
    describe('snapshot test',()=>{
        it('typeName==password',()=>{
            wrapper.setState({typeName:'password'})
            expect(wrapper).toMatchSnapshot()
        })
        it('typeName==phone',()=>{
            wrapper.setState({typeName:'email'})
            expect(wrapper).toMatchSnapshot()
        })

    })
   
})