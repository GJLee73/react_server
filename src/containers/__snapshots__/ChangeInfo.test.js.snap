// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeInfo test snapshot test typeName==password 1`] = `
<div>
  새로운 
  password
  을 입력하세요.
  <div
    className="card container"
  >
    <ChangeInfoPwd
      submit={[Function]}
    />
  </div>
</div>
`;

exports[`ChangeInfo test snapshot test typeName==phone 1`] = `
<div>
  새로운 
  email
  을 입력하세요.
  <div
    className="card container"
  >
    <ChangeInfoComponent
      submit={[Function]}
      typeName="email"
      value="g1"
    />
  </div>
</div>
`;
